{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Terminal commands\n",
    "mkdir lab7  \n",
    "cd lab7  \n",
    "new notebook Python3  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import packages\n",
    "import numpy as np\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "Entrez.email = 'y.peng@berkeley.edu'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.40000000000000013\n"
     ]
    }
   ],
   "source": [
    "# generate six random binary files that contain 50% to 100% of zeros; write to files\n",
    "prob = 1\n",
    "for i in range(6):\n",
    "    data = np.random.choice([0, 1], size=8*1024*1024*100, replace=True, p=[prob, 1-prob])\n",
    "    data = np.packbits(data)\n",
    "    open(\"zeros_\" + str(prob) + \"p\", \"wb\").write(data)\n",
    "    prob -= 0.1\n",
    "\n",
    "print(prob)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100000000"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# generate a random nucleotide sequence and a random amino acid sequence; write to files\n",
    "nt = np.random.choice([\"A\", \"T\", \"C\", \"G\"], size=100000000, replace=True)\n",
    "aa = np.random.choice(['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y'],\n",
    "                     size=100000000, replace=True)\n",
    "open(\"nt_seq.fa\", \"w\").write(\"\".join(nt))\n",
    "open(\"aa_seq.fa\", \"w\").write(\"\".join(aa))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Terminal commands\n",
    "### 100% zeros\n",
    "be131-20@meowth:~/lab7$ time gzip -k zeros_1p  \n",
    "  \n",
    "real    0m0.772s  \n",
    "user    0m0.760s  \n",
    "sys     0m0.012s  \n",
    "be131-20@meowth:~/lab7$ time bzip2 -k zeros_1p  \n",
    "  \n",
    "real    0m1.078s  \n",
    "user    0m1.022s  \n",
    "sys     0m0.056s  \n",
    "be131-20@meowth:~/lab7$ time pbzip2 -k -f zeros_1p  \n",
    "  \n",
    "real    0m0.112s  \n",
    "user    0m1.933s  \n",
    "sys     0m0.105s  \n",
    "be131-20@meowth:~/lab7$ time ArithmeticCompress zeros_1p zeros_1p.art  \n",
    "  \n",
    "real    0m18.691s  \n",
    "user    0m18.526s  \n",
    "sys     0m0.164s  \n",
    "### 90% zeros\n",
    "be131-20@meowth:~/lab7$ time gzip -k zeros_0.9p  \n",
    "  \n",
    "real    0m22.659s  \n",
    "user    0m22.586s  \n",
    "sys     0m0.073s  \n",
    "be131-20@meowth:~/lab7$ time bzip2 -k zeros_0.9p  \n",
    "  \n",
    "real    0m10.824s  \n",
    "user    0m10.735s  \n",
    "sys     0m0.089s  \n",
    "be131-20@meowth:~/lab7$ time pbzip2 -k -f zeros_0.9p  \n",
    "  \n",
    "real    0m0.769s  \n",
    "user    0m18.286s  \n",
    "sys     0m0.662s  \n",
    "be131-20@meowth:~/lab7$ time ArithmeticCompress zeros_0.9p zeros_0.9p.art  \n",
    "  \n",
    "real    0m28.733s  \n",
    "user    0m28.569s  \n",
    "sys     0m0.164s  \n",
    "### 80% zeros\n",
    "be131-20@meowth:~/lab7$ time gzip -k zeros_0.8p  \n",
    "  \n",
    "real    0m15.119s  \n",
    "user    0m14.911s  \n",
    "sys     0m0.156s  \n",
    "be131-20@meowth:~/lab7$ time bzip2 -k zeros_0.8p  \n",
    "  \n",
    "real    0m12.207s  \n",
    "user    0m12.107s  \n",
    "sys     0m0.100s  \n",
    "be131-20@meowth:~/lab7$ time pbzip2 -k -f zeros_0.8p  \n",
    "  \n",
    "real    0m0.959s  \n",
    "user    0m23.806s  \n",
    "sys     0m0.891s  \n",
    "be131-20@meowth:~/lab7$ time ArithmeticCompress zeros_0.8p zeros_0.8p.art  \n",
    "  \n",
    "real    0m35.441s  \n",
    "user    0m35.140s  \n",
    "sys     0m0.300s  \n",
    "### 70% zeros\n",
    "be131-20@meowth:~/lab7$ time gzip -k zeros_0.7p  \n",
    "  \n",
    "real    0m7.050s  \n",
    "user    0m6.893s  \n",
    "sys     0m0.156s  \n",
    "be131-20@meowth:~/lab7$ time bzip2 -k zeros_0.7p  \n",
    "  \n",
    "real    0m13.843s  \n",
    "user    0m13.690s  \n",
    "sys     0m0.153s  \n",
    "be131-20@meowth:~/lab7$ time pbzip2 -k -f zeros_0.7p  \n",
    "  \n",
    "real    0m1.184s  \n",
    "user    0m29.846s  \n",
    "sys     0m0.875s  \n",
    "be131-20@meowth:~/lab7$ time ArithmeticCompress zeros_0.7p zeros_0.7p.art  \n",
    "  \n",
    "real    0m39.259s  \n",
    "user    0m38.857s  \n",
    "sys     0m0.381s  \n",
    "### 60% zeros\n",
    "be131-20@meowth:~/lab7$ time gzip -k zeros_0.6p  \n",
    "  \n",
    "real    0m4.447s  \n",
    "user    0m4.315s  \n",
    "sys     0m0.133s  \n",
    "be131-20@meowth:~/lab7$ time bzip2 -k zeros_0.6p  \n",
    "  \n",
    "real    0m15.719s  \n",
    "user    0m15.547s  \n",
    "sys     0m0.173s  \n",
    "be131-20@meowth:~/lab7$ time pbzip2 -k -f zeros_0.6p  \n",
    "  \n",
    "real    0m1.439s  \n",
    "user    0m37.090s  \n",
    "sys     0m0.810s  \n",
    "be131-20@meowth:~/lab7$ time ArithmeticCompress zeros_0.6p zeros_0.6p.art  \n",
    "  \n",
    "real    0m41.033s  \n",
    "user    0m40.680s  \n",
    "sys     0m0.353s  \n",
    "### 50% zeros\n",
    "be131-20@meowth:~/lab7$ time gzip -k zeros_0.5p  \n",
    "  \n",
    "real    0m3.612s  \n",
    "user    0m3.464s  \n",
    "sys     0m0.149s  \n",
    "  \n",
    "be131-20@meowth:~/lab7$ time bzip2 -k zeros_0.5p  \n",
    "  \n",
    "real    0m18.525s  \n",
    "user    0m18.396s  \n",
    "sys     0m0.129s  \n",
    "be131-20@meowth:~/lab7$ time pbzip2 -k -f zeros_0.5p  \n",
    "  \n",
    "real    0m1.561s  \n",
    "user    0m40.698s  \n",
    "sys     0m0.951s  \n",
    "be131-20@meowth:~/lab7$ time ArithmeticCompress zeros_0.5p zeros_0.5p.art  \n",
    "  \n",
    "real    0m41.281s  \n",
    "user    0m40.885s  \n",
    "sys     0m0.344s  \n",
    "\n",
    "|   percentage   |   gzip   |   bzip2   |   pbzip2   |   art   |\n",
    "|----------------|----------|-----------|------------|---------|\n",
    "|      100%      |  102kB   |   113B    |   5.62kB   |  1.03kB |\n",
    "|      90%       |  58.7MB  |   61.2MB  |   61.2MB   |  49.2MB |\n",
    "|      80%       |  81.2MB  |   86.6MB  |   86.7MB   |  75.7MB |\n",
    "|      70%       |  93.6MB  |   99.8MB  |   99.8MB   |  92.4MB |\n",
    "|      60%       |  102MB   |   105MB   |   105MB    |  102MB  |\n",
    "|      50%       |  105MB   |   105MB   |   105MB    |  105Mb  |\n",
    "\n",
    "### Random nucleotide sequence\n",
    "be131-20@meowth:~/lab7$ time gzip -k nt_seq.fa  \n",
    "\n",
    "real    0m14.897s  \n",
    "user    0m14.772s  \n",
    "sys     0m0.124s  \n",
    "be131-20@meowth:~/lab7$ time bzip2 -k nt_seq.fa  \n",
    "\n",
    "real    0m9.474s  \n",
    "user    0m9.406s  \n",
    "sys     0m0.068s  \n",
    "be131-20@meowth:~/lab7$ time pbzip2 -k -f nt_seq.fa  \n",
    "\n",
    "real    0m0.674s  \n",
    "user    0m16.219s  \n",
    "sys     0m0.697s  \n",
    "be131-20@meowth:~/lab7$ time ArithmeticCompress nt_seq.fa nt_seq.art  \n",
    "\n",
    "real    0m21.333s  \n",
    "user    0m21.261s  \n",
    "sys     0m0.072s  \n",
    "\n",
    "### Random amino acid sequence\n",
    "be131-20@meowth:~/lab7$ time gzip -k aa_seq.fa  \n",
    "\n",
    "real    0m4.266s  \n",
    "user    0m4.166s  \n",
    "sys     0m0.100s  \n",
    "be131-20@meowth:~/lab7$ time bzip2 -k aa_seq.fa  \n",
    "\n",
    "real    0m12.695s  \n",
    "user    0m12.610s  \n",
    "sys     0m0.085s  \n",
    "be131-20@meowth:~/lab7$ time pbzip2 -k -f aa_seq.fa  \n",
    "\n",
    "real    0m0.816s  \n",
    "user    0m19.271s  \n",
    "sys     0m0.908s  \n",
    "be131-20@meowth:~/lab7$ time ArithmeticCompress aa_seq.fa aa_seq.art  \n",
    "\n",
    "real    0m28.631s  \n",
    "user    0m28.478s  \n",
    "sys     0m0.152s\n",
    "\n",
    "|    Sequence    |   gzip   |   bzip2   |   pbzip2   |   art   |\n",
    "|----------------|----------|-----------|------------|---------|\n",
    "|   Nucleotide   |  29.2MB  |   27.3MB  |   27.3MB   |   25MB  |\n",
    "|   Amino acid   |  60.6MB  |   55.3MB  |   55.3MB   |   54MB  |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Q: Which algorithm achieves the best level of compression on each file type?\n",
    "### A: ArithmeticCompress has the best level of compression on every file except zeros_1p. bzip2 is the best compression algorithm for zeros_1p.\n",
    "\n",
    "## Q: Which algorithm is the fastest?\n",
    "### A: pbzip2 is the fastest.\n",
    "\n",
    "## Q: What is the difference between bzip2 and pbzip2? Do you expect one to be faster and why?\n",
    "### A: pbzip2 is a parallel implementation of the bzip2 block-sorting file compressor that uses multiple cores in a processor. I expect pbzip2 to be faster.\n",
    "\n",
    "## Q: How does the level of compression change as the percentage of zeros increases? Why does this happen?\n",
    "### A: The level of compression increases as the percentage of zeros inceases. bzip2 and pbzip2 use Borrows-Wheeler transform to increase compression efficiency for the run-length encoding. Run-length encoding works by representing a series of zeros with two values: a zero followed by an integer that denotes the length of zeros.\n",
    "\n",
    "## Q: What is the minimum number of bits required to store a single DNA base?\n",
    "### A: 2 bits.\n",
    "\n",
    "## Q: What is the minimum number of bits required to store an amino acid letter?\n",
    "### A: 5 bits.\n",
    "\n",
    "## Q: In your tests, how many bits did gzip and bzip2 actually require to store your random DNA and protein sequences?\n",
    "### A: For the random DNA sequence, gzip uses 29.2MB and bzip2 uses 27.3 MB. For the random protein sequence, gzip uses 60.6MB and bzip2 uses 55.3MB.\n",
    "\n",
    "## Q: Are gzip and bzip2 performing well on DNA and proteins?\n",
    "### A: Theoretically, the 100Mbp DNA sequence can be compressed to 23.8MB. The 100Mbp protein sequence can be compressed to 59.6MB. gzip and bzip2 were not doing well on DNA but did very well on proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[SeqRecord(seq=Seq('ATGAGAGTGAGGGGGAGATTGAGGAATTGGCAACAATGGTGGATATGGATGTGG...CAA', SingleLetterAlphabet()), id='JQ362424.1', name='JQ362424.1', description='JQ362424.1 Human immunodeficiency virus clone TZBD.02 envelope glycoprotein (env) gene, partial cds', dbxrefs=[]), SeqRecord(seq=Seq('ATGAGAGTGAAGGGGATATTGAGGAATTGGCAACACAGGTGGATATGGATCTGG...CAA', SingleLetterAlphabet()), id='JQ362423.1', name='JQ362423.1', description='JQ362423.1 Human immunodeficiency virus clone TZA125.17 envelope glycoprotein (env) gene, partial cds', dbxrefs=[]), SeqRecord(seq=Seq('ATGAAAGTGAAGGAGACACAGATGAATTGGCTAAACTTGTGGAGATGGGGAACT...CTA', SingleLetterAlphabet()), id='JQ362422.1', name='JQ362422.1', description='JQ362422.1 Human immunodeficiency virus clone 620345.c1 envelope glycoprotein (env) gene, partial cds', dbxrefs=[]), SeqRecord(seq=Seq('ATGAGAGTGAGGGGGATACAGAGGAATTGTCAACAATGGTGGATATGGGGACTC...CTA', SingleLetterAlphabet()), id='JQ362421.1', name='JQ362421.1', description='JQ362421.1 Human immunodeficiency virus clone 288.38 envelope glycoprotein (env) gene, partial cds', dbxrefs=[]), SeqRecord(seq=Seq('ATGAAAGTGATGGGGATACCGAAGAATTGGCCACGATGGTGGATGTGGGGCATC...CAA', SingleLetterAlphabet()), id='JQ362420.1', name='JQ362420.1', description='JQ362420.1 Human immunodeficiency virus clone 286.36 envelope glycoprotein (env) gene, partial cds', dbxrefs=[]), SeqRecord(seq=Seq('ATGAGAGTGAAGGGGATGAGGAAGAATTATCAGCACTTGTGGAGATGGGGCACC...TAA', SingleLetterAlphabet()), id='KC595227.1', name='KC595227.1', description='KC595227.1 Human immunodeficiency virus isolate 10246788_14gp160 envelope glycoprotein (env) gene, complete cds', dbxrefs=[]), SeqRecord(seq=Seq('ATGAGAGTGATGGAGACCAGGAGGAATTATCAACACTTGTTGATATGGGGCATG...TAA', SingleLetterAlphabet()), id='KC595226.1', name='KC595226.1', description='KC595226.1 Human immunodeficiency virus isolate 10246788_3gp160 envelope glycoprotein (env) gene, complete cds', dbxrefs=[]), SeqRecord(seq=Seq('ATGAGAGTGATGGAGACCAGGAGGAATTATCAACACTTGTTGATATGGGGCATG...TAA', SingleLetterAlphabet()), id='KC595225.1', name='KC595225.1', description='KC595225.1 Human immunodeficiency virus isolate 10246788_1gp160 envelope glycoprotein (env) gene, complete cds', dbxrefs=[]), SeqRecord(seq=Seq('ATGAGAGTGAAGGAGATCAGGAAGAATTATCAGCATTGGTGGAAATGGGGCATC...TAA', SingleLetterAlphabet()), id='KC595224.1', name='KC595224.1', description='KC595224.1 Human immunodeficiency virus isolate 12_8_1_544_545 envelope glycoprotein (env) gene, complete cds', dbxrefs=[]), SeqRecord(seq=Seq('ATGAGAGTGAAGGGGATGAGGAAGAATTATCAGCACTTGTGGAAATGGGGCACC...TAA', SingleLetterAlphabet()), id='KC595223.1', name='KC595223.1', description='KC595223.1 Human immunodeficiency virus isolate 64.1-1gp160 envelope glycoprotein (env) gene, complete cds', dbxrefs=[])]\n"
     ]
    }
   ],
   "source": [
    "# Perform an Entrez search and fetch to acquire the FASTA sequences of \n",
    "# the envelope glycoprotein gp120 of HIV virus.\n",
    "seqs = []\n",
    "search = Entrez.esearch(db = 'nuccore',\n",
    "                        term = 'envelope glycoprotein AND \"Human immunodeficiency virus\"[porgn:__txid12721] AND 2000:5000[sequence length]',\n",
    "                        sort = 'relevance',\n",
    "                        idtype = 'acc',\n",
    "                        retstart = 0,\n",
    "                        retmax = 10)\n",
    "for i in Entrez.read(search)['IdList']:\n",
    "    gene_fetch = Entrez.efetch(db = 'nuccore', id = i, rettype = 'fasta', retmode = 'text')\n",
    "    record = SeqIO.read(gene_fetch, 'fasta')\n",
    "    seqs.append(record)\n",
    "\n",
    "with open(\"gp120.fasta\", \"w\") as output_handle:\n",
    "    SeqIO.write(seqs, output_handle, \"fasta\")\n",
    "    \n",
    "print(seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Q: A priori, do you expect to achieve better or worse compression here than random data? Why?\n",
    "### A: I expect worse compression with this multi-FASTA file because it contains one line of information about each isolates. This file is not purely consist of nucleotide sequences.\n",
    "\n",
    "|    File   |   gzip   |   bzip2   |   pbzip2   |   art   |\n",
    "|-----------|----------|-----------|------------|---------|\n",
    "|   gp120   |  5.55kB  |   5.96kB  |   5.96kB   | 9.13kB  |\n",
    "\n",
    "## Q: How does the compression ratio of this file compare to random data?\n",
    "### A: The compression ratio of gp120.fasta using gzip, bzip2, and pbzip2 is about 80%. The compression ratio of gp120.fasta using ArithmeticCompress is about 66%. The compression ratio of nt_seq.fa using gzip, bzip2, and pbzip2 is about 72%. The compression ratio of nt_seq.fa using ArithmeticCompress is about 75%. gzip, bzip2, and pbzip2 are better at compressing real nucleotide sequences. ArithmeticCompress is better at compressing randomly generated sequences."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Q: Given the benchmarking data you obtained in this lab, which algorithm do you propose to use for each type of data? Provide an estimate for the fraction of space you can save using your compression scheme. How much of a bonus do you anticipate receiving this year?\n",
    "\n",
    "### A: For the protein sequences and binary microscope images, I propose to use the ArithmeticCompress algorithm. For the genome and plasmid sequences, I propose to use the gzip algorithm.  I estimate that 50% of space can be saved for the protein sequence, close to 0 saving for the microscope images, and 80% saving for the genome and plasmid sequences. The total percentage of space saved would be $10\\% * 50\\% + 80\\% * 80\\% = 69\\%$. The amount of saved cost would be $69 * 500 = 34500$ dollars."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
